packages/contracts/contracts/delegateTargets/VirtualAppAgreement.sol

```diff
@@ -12,47 +12,49 @@ contract VirtualAppAgreement {
   using Transfer for Transfer.Terms;
 
   struct Agreement {
-    Registry registry;
-    NonceRegistry nonceRegistry;
+    address registryAddr;
     Transfer.Terms terms;
+    uint256 expiry;
+    bytes32 target;
     uint256 capitalProvided;
-    uint256 loanDuration;
-    bytes32 virtualStateDeposit;
-    address intermediary;
+    address[2] beneficiaries;
   }
 
   function delegateTarget(
-    bytes32 salt,
-    uint256 nonce,
     Agreement agreement
   )
     public
   {
+    Registry registry = Registry(agreement.registryAddr);
+    require(agreement.expiry <= block.number, "agreement has not expired yet");
+    address target = registry.resolver(agreement.target);
+    Transfer.Transaction memory resolution = AppInstance(target).getResolution();
 
-    require(
-      agreement.nonceRegistry.isFinalized(salt, nonce),
-      "VirtualAapp nonceRegistry was not finalized"
-    );
-
-    address[] memory to = new address[](2);
-    to[0] = agreement.registry.resolver(agreement.virtualStateDeposit);
-    to[1] = agreement.intermediary;
+    require(resolution.value.length == 1, "returned invalid resolution");
+    require(resolution.to.length == 1, "returned invalid resolution");
+    require(agreement.terms.assetType == resolution.assetType, "returned incompatible resolution");
+    require(agreement.terms.token == resolution.token, "returned incompatible resolution");
+    require(agreement.capitalProvided > resolution.value[0], "returned incompatible resolution");
```

Comment:
The logic for all of this is pretty critical. I'd like to recommend we document this in the header of the file, or perhaps on a docstring for the function. Ideally we reference to it on specs.counterfactual.com.